## A. Στοιχεία Ομάδας

Αριθμός Ομάδας: Μ.Α.11

> Καρυπιάδης Ευστράτιος: 03114911 

---

## Β. Εισαγωγή του Dataset

Το dataset μας προκύπτει έπειτα απο την εφαρμογή του συστήματος _HIPS (Hierarchical Image Processing System)_ σε σειρά εικόνων οχημάτων, οι οποίες ελήφθησαν για 4 διαφορετικά οχημάτα και κάτω από τις ίδιες συνθήκες φωτισμού και γωνίας λήψης. Τα χαρακτηριστικά που εξήχθησαν αποτελούν και τις στήλες των δεδομένων μας, βάσει των οποίων θα επιχειρήσουμε να ταξινομήσουμε τα δείγματα που έχουμε στην διάθεσή μας στην αντίστοιχη κατηγορία που ανήκουν.

Τα αρχεία δεδομένων μας είναι χωρισμένα σε 9 διαφορετικά αρχεία τα οποία θα ενώσουμε σε ένα _DataFrame_. Δεν υπάρχει αρίθμηση για κάθε διαφορετικό δείγμα - γραμμή, αλλά ούτε ονομασία για τα χαρακτηριστικά μας. Για αυτό το λόγο θα ονοματίσουμε της στήλες των δεδομένων μας, συμβουλεύομενοι την περιγραφή του dataset.

In [1]:
import glob
import pandas as pd

path = 'Data'
all_data = glob.glob(path + "/*.dat")

l = []

column_names = ['compactness', 'circularity', 'distance_circularity', 'radius_ratio', 'pr.axis_aspect_ratio',
                'max.length_aspect_ratio', 'scatter_ratio', 'elongatedness', 'pr.axis_rectangularity', 
                'max.length_rectangularity', 'major.axis_scaled_variance', 'minor.axis_scaled_variance',
                'gyration_scaled_radius', 'major.axis skewness', 'minor.axis_skewness', 'minor.axis_kurtosis',
                'major.axis_curtosis', 'hollows_ratio', 'label']

for filename in all_data:
    df = pd.read_csv(filename, header=None, sep=' ', names = column_names, usecols= [num for num in range(19)])
    l.append(df)
    
cars = pd.concat(l, axis=0, ignore_index=True)
cars.head(10)

compactness  circularity  distance_circularity  radius_ratio  \
0           95           48                    83           178   
1           91           41                    84           141   
2          104           50                   106           209   
3           93           41                    82           159   
4           85           44                    70           205   
5          107           57                   106           172   
6           97           43                    73           173   
7           90           43                    66           157   
8           86           34                    62           140   
9           93           44                    98           197   

   pr.axis_aspect_ratio  max.length_aspect_ratio  scatter_ratio  \
0                    72                       10            162   
1                    57                        9            149   
2                    66                       10            207   
3                    63                        9            144   
4                   103                       52            149   
5                    50                        6            255   
6                    65                        6            153   
7                    65                        9            137   
8                    61                        7            122   
9                    62                       11            183   

   elongatedness  pr.axis_rectangularity  max.length_rectangularity  \
0             42                      20                        159   
1             45                      19                        143   
2             32                      23                        158   
3             46                      19                        143   
4             45                      19                        144   
5             26                      28                        169   
6             42                      19                        143   
7             48                      18                        146   
8             54                      17                        127   
9             36                      22                        146   

   major.axis_scaled_variance  minor.axis_scaled_variance  \
0                         176                         379   
1                         170                         330   
2                         223                         635   
3                         160                         309   
4                         241                         325   
5                         280                         957   
6                         176                         361   
7                         162                         281   
8                         141                         223   
9                         202                         505   

   gyration_scaled_radius  major.axis skewness  minor.axis_skewness  \
0                     184                   70                    6   
1                     158                   72                    9   
2                     220                   73                   14   
3                     127                   63                    6   
4                     188                  127                    9   
5                     264                   85                    5   
6                     172                   66                   13   
7                     164                   67                    3   
8                     112                   64                    2   
9                     152                   64                    4   

   minor.axis_kurtosis  major.axis_curtosis  hollows_ratio label  
0                   16                  187            197   van  
1                   14                  189            199   van  
2                    9                  188            196  saab  
3                   10 

Στη συνέχεια θα εξάγουμε χρήσιμες πληροφορίες για τα δεδομένα μας.

In [2]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 846 entries, 0 to 845
Data columns (total 19 columns):
compactness                   846 non-null int64
circularity                   846 non-null int64
distance_circularity          846 non-null int64
radius_ratio                  846 non-null int64
pr.axis_aspect_ratio          846 non-null int64
max.length_aspect_ratio       846 non-null int64
scatter_ratio                 846 non-null int64
elongatedness                 846 non-null int64
pr.axis_rectangularity        846 non-null int64
max.length_rectangularity     846 non-null int64
major.axis_scaled_variance    846 non-null int64
minor.axis_scaled_variance    846 non-null int64
gyration_scaled_radius        846 non-null int64
major.axis skewness           846 non-null int64
minor.axis_skewness           846 non-null int64
minor.axis_kurtosis           846 non-null int64
major.axis_curtosis           846 non-null int64
hollows_ratio                 846 non-null int64
label        

In [12]:
cars.describe()

compactness  circularity  distance_circularity  radius_ratio  \
count   846.000000   846.000000            846.000000    846.000000   
mean     93.678487    44.861702             82.088652    168.940898   
std       8.234474     6.169866             15.771533     33.472183   
min      73.000000    33.000000             40.000000    104.000000   
25%      87.000000    40.000000             70.000000    141.000000   
50%      93.000000    44.000000             80.000000    167.000000   
75%     100.000000    49.000000             98.000000    195.000000   
max     119.000000    59.000000            112.000000    333.000000   

       pr.axis_aspect_ratio  max.length_aspect_ratio  scatter_ratio  \
count            846.000000               846.000000     846.000000   
mean              61.693853                 8.567376     168.839243   
std                7.888251                 4.601217      33.244978   
min               47.000000                 2.000000     112.000000   
25%               57.000000                 7.000000     146.250000   
50%               61.000000                 8.000000     157.000000   
75%               65.000000                10.000000     198.000000   
max              138.000000                55.000000     265.000000   

       elongatedness  pr.axis_rectangularity  max.length_rectangularity  \
count     846.000000              846.000000                 846.000000   
mean       40.933806               20.582742                 147.998818   
std         7.811560                2.592138                  14.515652   
min        26.000000               17.000000                 118.000000   
25%        33.000000               19.000000                 137.000000   
50%        43.000000               20.000000                 146.000000   
75%        46.000000               23.000000                 159.000000   
max        61.000000               29.000000                 188.000000   

       major.axis_scaled_variance  minor.axis_scaled_variance  \
count                  846.000000                  846.000000   
mean                   188.625296                  439.911348   
std                     31.394837                  176.692614   
min                    130.000000                  184.000000   
25%                    167.000000                  318.250000   
50%                    178.500000                  364.000000   
75%                    217.000000                  587.000000   
max                    320.000000                 1018.000000   

       gyration_scaled_radius  major.axis skewness  minor.axis_skewness  \
count               846.00000           846.000000           846.000000   
mean                174.70331            72.462175             6.377069   
std                  32.54649             7.486974             4.918353   
min                 109.00000            59.000000             0.000000   
25%                 149.00000            67.000000             2.000000   
50%                 173.00000            71.500000             6.000000   
75%                 198.00000            75.000000             9.000000   
max                 268.00000           135.000000            22.000000   

       minor.axis_kurtosis  major.axis_curtosis  hollows_ratio  
count           846.000000           846.000000     846.000000  
mean             12.599291           188.932624     195.632388  
std               8.931240             6.163949       7.438797  
min               0.000000           176.000000     181.000000  
25%               5.000000           184.000000     190.250000  
50%              11.000000           188.000000     197.000000  
75%              19.000000           193.000000     201.000000  
max              41.000000           206.000000     211.000000

Παρατηρούμε ότι έχουμε στη διάθεσή μας 846 δείγματα, καθένα από τα οποία συνοδεύεται από 18 συνολικά χαρακτηριστικά. Τα χαρακτηριστικά αυτά είναι αριθμοί, ενώ δεν υπάρχει στήλη στην οποία να απουσιάζουν τιμές. Η τελευταία στήλη αντιστοιχεί στις ετικέτες των δεδομένων μας, οι οποίες ωστόσο είναι σε μορφή κειμένου και τις οποίες δε χρειάζεται να κωδικοποιήσουμε.

In [3]:
cars['label'].value_counts(normalize=True) * 100

bus     25.768322
saab    25.650118
opel    25.059102
van     23.522459
Name: label, dtype: float64

Σημειώνουμε ότι οι κατηγορίες των οχημάτων μας είναι 4 και φέρουν τα ονόματα: _Bus, Van, Opel_ και _Saab_. Έχουμε ισορροπία στο σύνολο των δεδομένων, αφού όπως υπολογίζουμε κάθε κατηγορία συναστά περίπου το 1/4 του dataset.

Αν και οι κατηγορίες μας είναι ισορροπημένες στο σύνολο των δεδομένων μας, για το διαχωρισμό αυτών θα χρησιμοποιήσουμε τη _StratifiedShuffleSplit_ ώστε να εξασφαλίσουμε ότι τα δεδομένα εκπαίδευσης θα αποτελούνται από αντιπροσωπευτικό πλήθος κατηγοριών με το ολικό dataset.

In [8]:
import numpy as np
from sklearn.model_selection import StratifiedShuffleSplit

split = StratifiedShuffleSplit(n_splits = 1, test_size = 0.2, random_state = 42)
for train_index, test_index in split.split(cars, cars["label"]):
    strat_train_set = cars.reindex(train_index)
    strat_test_set = cars.reindex(test_index)
    
strat_train_set["label"].value_counts() / len(strat_train_set) * 100

bus     25.739645
saab    25.591716
opel    25.147929
van     23.520710
Name: label, dtype: float64

___

## Γ. Baseline Classification